Add a new episode dealing with CMORizers and observational data

Peter9192 opened the pull request:
# Pull Request checklist

We appreciate your time and effort to improve the tutorial. Please keep in mind that lesson maintainers are volunteers and it may be some time before they can respond to your contribution.

* * *

## Before you start

- [x] Read [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md).
- [x] Create an [issue](https://github.com/ESMValGroup/tutorial/issues) to discuss your idea. This allows your contributions to be incorporated into the tutorial.

## Tasks

- [x] Give this pull request a descriptive title.
- [ ] If you are contributing to existing lesson materials, please make sure the content conforms to the `Lesson development` section in [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md) and does not contain any spelling or grammatical errors.
- [ ] If you are making a new episode, please make sure the content conforms to the `Lesson organization` and `Lesson formatting` sections in [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md) and does not contain any spelling or grammatical errors.
- [ ] Preferably Codacy checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why.
- [ ] Preview changes on your machine before pushing them to GitHub by running `make serve`, alternatively `make docker-serve`. Please see the `Previewing your changes locally` section in [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md) for installation instructions.
- [ ] All code instructions have been tested.

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes #168

Peter9192 commented on _episodes/09-cmorization.md:
````diff
@@ -0,0 +1,1046 @@
+> > ## Answers
+> >
+> > ```bash
+> > esmvaltool run recipe_check_fluxnet.yml --log_level debug
````
I think it would still not find the data at this point since we have not yet run the cmorizer, right? So all we have is RAWOBS, whereas we need some data in OBS6 to get this error.

Maybe we still need to move the section "## Check if your variable is following the CMOR standard / Check if it's in a CMOR table" further down, so that the first thing we do is run the (almost) empty cmorizer script.

hb326 replied:
@Peter9192: yes, you are right. The ESMValTool would not find the data yet. It just did for me, I remember now, because I had it look in the RAWOBS folder...
And yes, I agree, we probably should have the "CMOR standard section" a little further down. Or at least have the people try to run an almost empty cmorizer script after the section "Location, location, location".

hb326 replied:
@Peter9192: actually, I have thought about a different approach, how to adjust our story line so that it makes more sense. I have implemented that now. Let me know what you think!

Peter9192 commented:
@hb326 nice work, I think it's fairly complete as it is now! I'm seeing a few things that I think we can still optimize a bit, but maybe at this point it would be good to have some other people look at it first

hb326 commented:
> 
> 
> @hb326 nice work, I think it's fairly complete as it is now! I'm seeing a few things that I think we can still optimize a bit, but maybe at this point it would be good to have some other people look at it first

@Peter9192 I agree. I can also think about a few things that might need some polishing and some minor adjusting, but I think now it is kind of at the time to have fresh sets of eyes look at the episode.

Peter9192 commented:
@SarahAlidoost would you have time to do a review? I think if the development installation episode you worked on comes before this one, we can refer back to that at some places. 

Like we discussed earlier, we should also have some external users tests it once it's merged.

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+---
+title: "CMORization: Using observational datasets"
```
```suggestion
title: "CMORization: using observational datasets"
```

Peter9192 replied:
I'll change the title (and take care of the capitals)

Peter9192 replied:
Done

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+questions:
+- "What is so challenging about observational data?"
+- "How do I use observational datasets in ESMValTool?"
+- "How add support for new (observational) datasets?"
```
```suggestion
- "How do I add support for new (observational) datasets?"
```

Peter9192 replied:
Changed to "how to add"

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+
+## Introduction
+
+This episode deals with "CMORization". ESMValTool was designed to work with data
```
```suggestion
This episode deals with "CMORization". ESMValTool is designed to work with data
```

Peter9192 replied:
Done

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+coordinate information, how the data should be structured (e.g. 1 variable per
+file), additional metadata requirements, but also file naming conventions a.k.a.
+the data reference syntax (DRS). All this information is stored in so-called
+CMOR tables. As example, the CMOR tables for the CMIP6 project can be found
```
```suggestion
CMOR tables. As an example, the CMOR tables for the CMIP6 project can be found
```

Peter9192 replied:
Done

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+go through all the steps and explain relevant topics as we go. If you prefer to
+implement CMOR fixes, please read the documentation
+[here](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/develop/fixing_data.html#fixing-data).
+While fixes are implemented slightly differently, conceptually the process is
```
Please re-phrase this sentence regarding the usage of these three words "slightly differently, conceptually". The readability level of this sentence is quite low.

Peter9192 replied:
Rephrased completely

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+> button on the "FLUXCOM (RS+METEO) Global Land Carbon Fluxes using CRUNCEP
+> climate data". You'll be send an FTP address to access the server. Connect
+> to the server, follow the path in your email, and look for the file
+> `raw/monthly/GPP.ANN.CRUNCEPv6.monthly.2000.nc`. Download that file and save in
```
```suggestion
> `raw/monthly/GPP.ANN.CRUNCEPv6.monthly.2000.nc`. Download that file and save
```

Peter9192 replied:
Corrected

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+> climate data". You'll be send an FTP address to access the server. Connect
+> to the server, follow the path in your email, and look for the file
+> `raw/monthly/GPP.ANN.CRUNCEPv6.monthly.2000.nc`. Download that file and save in
+> in a folder called `/RAWOBS/Tier3/`.
```
The `/RAWOBS/Tier3/` is a directory rather than a folder name here. Can we make it clear by re-phrasing it?

Peter9192 replied:
Not sure I understand. What's the difference between a directory and a folder name?

SarahAlidoost replied:
Because the folder name is `FLUXCOM` here, I thought the sentence “save it in a folder called `/RAWOBS/Tier3/FLUXCOM` " might be confusing from a user’s  (researcher) perspective as `/RAWOBS/Tier3/FLUXCOM ` shows three folders. However, from a developer’s perspective, I am aware that a folder is identified in this way. I see several cases in the lesson that a file or a folder is called in this way. So, it is fine with me.

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+> in a folder called `/RAWOBS/Tier3/`.
+>
+> Note: you'll need a user-friendly ftp client. On Linux, `ncftp` works okay.
+{: .challenge}
```
Why the instruction about getting the data should be in a {:challenge} box while no solution is provided for this challenge? Also, we first teach something and then design a challenge. So users can practice what they have already learned. I suggest making the instruction plain text and remove one of these titles `Obtaining the data` and `Get the data`.

Peter9192 replied:
Sorry I thought the challenge was also meant to distinguish exercise from theory. I've changed it to plain text.

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+datasets are included in ESMValTool, and ESMValTool also provides a convenient
+way to execute them.
+
+Occasionally it happens that there are still minor issue with CMIP datasets. In
```
```suggestion
Occasionally it happens that there are still minor issues with CMIP datasets. In
```

Peter9192 replied:
I removed this sentence altogether

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+> register. After registration, in the dropdown boxes, select FLUXCOM as the data
+> choice and click download. Three files will be displayed. Click the download
+> button on the "FLUXCOM (RS+METEO) Global Land Carbon Fluxes using CRUNCEP
+> climate data". You'll be send an FTP address to access the server. Connect
```
```suggestion
> climate data". You'll receive an FTP address to access the server. Connect
```

Peter9192 replied:
Done

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+> Create a simple recipe that loads the "FLUXCOM" data. It should include a
+> datasets section with a single entry for the "FLUXCOM" dataset with the correct
+> dataset keys, and a diagnostics section with two variables: gpp and gppStderr.
+> We won't need any preprocessors or scripts (set `scripts: null`), but you will
```
```suggestion
> We don't need any preprocessors or scripts (set `scripts: null`), but we
```

Peter9192 replied:
done

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+> >
+> > Note: a recipe similar to this one is available under
+> > `~/path/to/ESMValTool/esmvaltool/recipes/examples/recipe_check_obs.yml`.
+> > That recipe includes checks all datasets for which CMORizers are available.
```
I suggest replacing the path with its link e.g. [recipe_check_obs ](https://github.com/ESMValGroup/ESMValTool/blob/master/esmvaltool/recipes/examples/recipe_check_obs.yml.). Why users need to know about this recipe? Can we provide more information by referring users to some ESMValTool documentations? 

Another issue, we can add a tip here guiding users to have a look at the episode `Writing your worn recipe`.

Peter9192 replied:
Removed the sentence with recipe_check_obs, and added the link to the episode on writing your own recipe 👍

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+
+> ## Create a test recipe
+>
+> Create a simple recipe that loads the "FLUXCOM" data. It should include a
```
```suggestion
> Create a simple recipe name `recipe_check_fluxnet.yml ` that loads the "FLUXCOM" data. It should include a
```

Peter9192 replied:
Done

SarahAlidoost commented on _episodes/09-cmorization.md:
````diff
@@ -0,0 +1,1053 @@
+> {: .solution}
+{: .challenge}
+
+Try to run the example recipe with
+
+```bash
+esmvaltool run recipe_check_fluxcom.yml --log_level debug
+```
+
+The `log_level` flag ensures that all relevant information is included in the
+output. We will need this information to find out what needs to be done. Look
+carefully through the log messages. You'll probably find something like
+
+```
+DEBUG   Retrieving OBS6 configuration
+DEBUG   Skipping non-existent /home/peter/default_inputpath/Tier3/FLUXCOM
+DEBUG   Looking for files matching ['OBS6_FLUXCOM_reanaly_ANN-v1_Lmon_gppStderr[_.]*nc'] in []
+ERROR   No input files found for variable {'variable_group': 'gppStderr' ...
+...
+ERROR   Missing data for preprocessor FLUXCOM/gppStderr
+...
+ERROR   Program terminated abnormally, see stack trace below for more information:
+...
+```
+{: .error}
+
+So the output tells us that it cannot find the data, and also where it has been
+looking. Let's try to understand what's going on.
+
+## Location, location, location
+
+The error messages above tell you that ESMValTool cannot find the data. That
+makes total sense, because we have not yet created our CMORized copy of the
````
Sorry, I am lost here. I expect this episode shows me how to create a cmorizer script. There are already three well-written episodes on Configuration (episode 3), writing your own recipe (episode 5) and debugging (episode 6). Please avoided duplications and just point the users to those lessons. I see that the debugging exercise here is a nice one, please add it to episode 6.

Peter9192 replied:
The idea was that the debug messages tell us what we need to fix for the cmorizer. How else would you know what kind of fixes the cmorizer needs to implement? But I can see why you're confused, and I think I know a way to avoid this here.

Peter9192 replied:
I removed this, but the episode still uses the error messages in a few places. Is it okay like this, or do you still think it's confusing?

SarahAlidoost replied:
It is much clear now. Thank you for addressing my comment.

SarahAlidoost commented on _episodes/09-cmorization.md:
````diff
@@ -0,0 +1,1053 @@
+>   OBS6: /path/to/my/obs6/data
+>   RAWOBS: /path/to/my/rawobs/data
+> ```
+{: .challenge}
````
Why the instruction about setting the correct path is in a {:challenge} box while no solution is provided for this challenge? Also, we first teach something and then design a challenge. So users can practice what they have already learned. I suggest making the instruction plain text.

Peter9192 replied:
Sorry I thought the challenge was also meant to distinguish exercise from theory. I've changed it to plain text.

SarahAlidoost replied:
It is right that a challenge is "meant to distinguish exercise from theory". However, setting the correct path (here) and downloading data (in the above) are more **instructions** rather than **exercises**. 
A challenge provides learners an opportunity to practice what that you’ve included in the skill list (objectives) for designing the lesson. In designing a challenge, we have to first teach something then challenge learners if they can do it. I hope now it is clear why I asked to convert those instructions to plain texts.

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+>
+{: .callout}
+
+Looking closer at the path that ESMValTool told us it went looking for the data,
+we see that it has a very specific structure:
+
+`OBS6_FLUXCOM_reanaly_ANN-v1_Lmon_gppStderr[_.]*nc`
+
+Note that all components in the filename correspond to one of the dataset keys
+defined in our test recipe. By replacing them with the original tags, we can
+see the general structure for the OBS6 data:
+
+`[project]_[dataset]_[type]_[version]_[mip]_[short_name]_*.nc`
+
+This is the Data Reference Syntax (DRS) for OBS6 data. Each project has its own
+DRS, and some project IDs even have multiple different options. These settings
+can be found in the file
+[config-developer.yml](https://github.com/ESMValGroup/ESMValCore/blob/master/esmvalcore/config-developer.yml).
+Luckily, we don't need to worry about that. This file name will be automatically
+correctly created when we run our CMORizer script.
```
In this part, I am introduced to an example of DRS and then to the definition of DRS. Can we reverse the order i.e. first explain what a DRS is and then show me an example and then a reference to config-developer? 

Another point, in episode **configuration**, there is a section **Directory structure for the data from different projects** that explains DRS. Can we point users to that section?

Peter9192 replied:
I removed most of this.

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+those cases, it is possible to fix those issues in ESMValCore before any further
+processing is done. The same can be done for non-CMIP data. The advantage is
+that you don't need to store an additional, reformatted copy of the data. The
+disadvantage is that these fixes should be implemented inside ESMValCore.
```
Please add a note here that this tutorial does not cover the development or any lessons about how to use esmvalcore.

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+
+
+If the variable you are interested in is not available in the standard CMOR tables,
+you need to write a custom CMOR table entry for the variable. Don't worry! It sounds
+more complicated than it is! Examples of custom CMOR table entries are for example
```
```suggestion
you need to write a custom CMOR table entry for the variable. Examples of custom CMOR table entries are for example
```

Peter9192 replied:
This is no longer part of the episode

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+if the data is actually following the necessary CMOR standard already,
+or if we have to do some reformatting for the dataset.
+
+> ## Run the test recipe again
```
These sections are not shown correctly on the tutorial webpage. Please preview the lesson locally and fix it.

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+The first step now is to create a file in the right folder that will contain
+the short python script. The home of all CMORizer scripts for observations
+and reanalysis datasets is
+[here](https://github.com/ESMValGroup/ESMValTool/tree/master/esmvaltool/cmorizers/obs).
```
I suggest replacing `[here] `to `esmvaltool/cmorizers/obs` and point users to episode development-setup where they can ding these directories.

Peter9192 replied:
Good point. I'll instead add a note about the development setup episode near the top of the episode.

Peter9192 replied:
I added several references to the dev setup episode, and I now point to these paths explicitly

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+> ## Note
+>
+> Always, always, when modifying or creating new code for the ESMValTool
+> repositories, work on your *own, local* branch of the ESMValTool. Optimally,
+> you have forked that branch directly from the most up-to-date version of
+> the "master" branch to avoid conflicts later when you want to merge your
+> code with the "master" branch of the ESMValTool.
```
```suggestion
> repositories, work on your *own, local* branch of the ESMValTool. For more information see lesson [Development and contribution]({{ page.root }}{% link _episodes/08-development-setup.md %}).
```

Peter9192 replied:
Pointing to the dev episode now

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+the last time, which ESMValTool "tier" it is associated with, and more
+detailed information about the necessary downloading and processing steps.
+
+> ## Fill out the header for the "FLUXCOM" dataset
```
These sections are not shown correctly on the tutorial webpage. Please preview the lesson locally and fix it.

SarahAlidoost commented on _episodes/09-cmorization.md:
````diff
@@ -0,0 +1,1053 @@
+
+Therefore it is necessary to create a configuration file for the new dataset.
+This configuration file needs to be stored in the
+following folder:
+
+```bash
+ESMValTool/esmvaltool/cmorizers/obs/cmor_config/
+```
````
```suggestion
following folder: ``ESMValTool/esmvaltool/cmorizers/obs/cmor_config/``
```

Peter9192 replied:
No long part of the episode

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+> >
+> > The original configuration file for the "FLUXCOM" dataset can be
+> > found here:
+> > [FLUXCOM.yml]
+> > (https://github.com/ESMValGroup/ESMValTool/blob/master/esmvaltool/cmorizers/obs/cmor_config/FLUXCOM.yml)
```
```suggestion
> > [FLUXCOM.yml](https://github.com/ESMValGroup/ESMValTool/blob/master/esmvaltool/cmorizers/obs/cmor_config/FLUXCOM.yml)
```

Peter9192 replied:
Done

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+> > Note the attribute "reference" here: it should include a ``doi`` related
+> > to the dataset. For more information on how to add references to the
+> > ``reference`` section of the configuration file, see the section in the
+> > documentation about this: [adding references]
+> > (https://docs.esmvaltool.org/en/latest/community/diagnostic.html#adding-references)
```
```suggestion
> > documentation about this: [adding references](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#adding-references)
```

Peter9192 replied:
No longer part of the episode

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+> > documentation about this: [adding references]
+> > (https://docs.esmvaltool.org/en/latest/community/diagnostic.html#adding-references)
+> >
+> > If a single dataset has more than one reference, it is possible to add
+> > tags as a list e.g. ``reference: ['tag1', 'tag2']``.
```
This information is already mentioned in the documentation and it beyond the scope of this episode. Let's remove it to make the lesson concise.

Peter9192 replied:
Done

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+the  ``.yml`` configuration file. The return value of this function is ignored.
+All the work, i.e. loading of the raw files, processing them and saving the
+final output, has to be performed inside its body. To simplify this process,
+ESMValTool provides a set of predefined utilities.py_, which can be imported
```
```suggestion
ESMValTool provides a set of functions in ``utilities.py``, which can be imported
```

Peter9192 replied:
No longer present

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+filepath to the correct one and the function ``_extract_variable`` extracts and
+saves a single variable from the raw data.
+
+After all that theory, let's have a look at the actualy python code of the
```
```suggestion
After all that theory, let's have a look at the python code of the
```

Peter9192 replied:
No longer part of the episode

SarahAlidoost commented on _episodes/09-cmorization.md:
````diff
@@ -0,0 +1,1053 @@
+```
+
+Let's run this CMORizing script to see if the dataset is read correctly, and
+what kind of file is written out. Tere is a specific command available in the
````
```suggestion
what kind of file is written out. There is a specific command available in the
```

Peter9192 replied:
Done

SarahAlidoost commented on _episodes/09-cmorization.md:
````diff
@@ -0,0 +1,1053 @@
+> ## Was the CMORization successful so far?!
+>
+> If you check the folders in your output path, you should see the following
+> folder structure:
+> ```bash
+> /Tier3/FLUXCOM/
+> ```
````
```suggestion
> folder structure: ``/Tier3/FLUXCOM/``
```

Peter9192 replied:
No longer part of the episode

SarahAlidoost commented on _episodes/09-cmorization.md:
````diff
@@ -0,0 +1,1053 @@
+> /Tier3/FLUXCOM/
+> ```
+>
+> Within the "FLUXCOM" folder there should be a NetCDF file with the name:
+> ```bash
+> OBS_FLUXCOM_reanaly_ANN-v1_Lmon_gpp_xxxx01-xxxx12.nc
+> ```
````
```suggestion
> Within the "FLUXCOM" folder there should be a NetCDF file with the name:
> ``OBS_FLUXCOM_reanaly_ANN-v1_Lmon_gpp_xxxx01-xxxx12.nc``
```

Peter9192 replied:
done

SarahAlidoost commented on _episodes/09-cmorization.md:
````diff
@@ -0,0 +1,1053 @@
+cube.units = 'kg m-2 s-1'
+```
+
+The whole section should not look then look like this:
````
```suggestion
The whole section should then look like this:
```

Peter9192 replied:
No longer present

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+relatively often for coordinates like ``lat``, ``lon`` or ``time``. This means
+that these problems need fixing in many CMORizers. Therefore there are common
+functions available within the ESMValTool that one can import and use in the
+new CMORizer script. The functions, written in python, are stored in the folder
+[utilities.py](https://github.com/ESMValGroup/ESMValTool/blob/master/esmvaltool/cmorizers/obs/utilities.py)
```
The script utilities.py has already been introduced above.

Peter9192 replied:
Restructured so this is different now

SarahAlidoost commented on _episodes/09-cmorization.md:
````diff
@@ -0,0 +1,1053 @@
+>         double lon_bnds(lon, bnds) ;
+> ```
+>
+> For that to happen, you will have to fix/work on the following things:
````
```suggestion
> For that to happen, you have to fix/work on the following things:
```

Peter9192 replied:
No longer present

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+the ESMValTool can work with it, it would be great if you could provide the
+CMORizer, and ultimately with that the dataset, to the rest of the community.
+To do that there are a few more steps you have to do:
+1. Open a pull request in the ESMValTool repository describing the dataset briefly
```
We encourage users to discuss their ideas in an issue as the first step of contribution. So, I suggest replacing this step with a sentence pointing to the episode 08-development-setup where the contribution procedure is explained.

Peter9192 replied:
Done

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+3. Make sure that there is a reference file available for the dataset [BibTeX info file](https://github.com/ESMValGroup/ESMValTool/tree/master/esmvaltool/references)
+
+More information about working with pull requests are available in the ESMValTool
+documentation under [Contributing a review](https://esmvaltool--1920.org.readthedocs.build/en/1920/community/review.html)
```
```suggestion
documentation under [Contributing a review](https://esmvaltool--1920.org.readthedocs.build/en/1920/community/review.html).
```

Peter9192 replied:
Done

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+## Some final comments
+
+Adding a new CMORizer to the ESMValTool is definitely already an advanced task
+when working with the ESMValTool. You have to have a basic understanding of
```
```suggestion
when working with the ESMValTool. You need to have a basic understanding of
```

Peter9192 replied:
Done

SarahAlidoost commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,1053 @@
+of the ESMValTool and ESMValCore.
+
+More information about adding observations to the ESMValTool can be found in the
+[documentation](https://docs.esmvaltool.org/en/latest/input.html#observations)
```
```suggestion
[documentation](https://docs.esmvaltool.org/en/latest/input.html#observations).
```

Peter9192 replied:
Done

SarahAlidoost reviewed (changes requested):
@Peter9192 and @hb326 thank you for the effort and nice work. I added some comments on the pull request. Here, my major concerns are:

- I like the content in the section “Create a CMORizer for the new dataset”. It is the main part of this lesson. Pity that users need to scroll down a lot to reach this part. Can we summarize the first sections before this section? Also, this section is a very long section. It would be great if you can organize it into sub-sections. 
- Can we use a different dataset in this episode that does not need users to register? Then we can use this lesson in any future training activities. 
- Please use a variable that its custom table is already available in esmvalcore. I believe that guiding users to create a custom table in esmvalcore is beyond the scope of this tutorial. Because the ESMVAlTool tutorial is meant to cover topics using esmvaltool rather esmvalcore. If we want to develop lessons about esmvalcore, I suggest discussing this in one of our meetings.

Peter9192 commented:
Hey @SarahAlidoost thanks for the thorough review! @hb326 I have some time to work on this tomorrow, shall I have a first go at responding to Sarahs comments?

hb326 commented:
> 
> 
> Hey @SarahAlidoost thanks for the thorough review! @hb326 I have some time to work on this tomorrow, shall I have a first go at responding to Sarahs comments?

@Peter9192, awesome if you will give it a try tomorrow. Let me know when you finished for the day or get stuck with something.

Peter9192 commented:
Hey @hb326 I have started on restructuring the episode, but it's been tougher than I thought. I'm not done yet, but if you're looking at this right now and don't like where it's going, better stop me now before I put any more effort into it. No hard feelings in that case ;-)

hb326 commented:
@Peter9192: I just checked the current structure of the episode, with the changes you have made so far. I think I understand your  suggested new structure, and I like it. So there are definitely no objections.  :) 
Let me know if you need any input from me. For now I will not touch the episode so that you can keep working on the restructuring.

Peter9192 commented:
Hi @hb326 I'm almost done with restructuring the episode. Most important changes:

- I put the section on 'running the existing cmorizer' first, so people will know what they are working towards.
- I'm no longer trying to work towards the old CMORizer script. I think the new approach/implementation is simpler. 
- Some parts have been left out. For example, the custom CMOR table exercise has been replaced with a comment that this is possible, but beyond the scope. 
- I put all the 'not strictly necessary' steps at the bottom. E.g. adding a header to a cmorizer script. This helps to put more focus and emphasis on what the purpose of each line of code in the cmorizer is. 

The work is not done yet. I got as far as subsection 4 (finalizing the cmorizer). Would you have time to go over what I have so far, and maybe work a little bit more on the last part? Some of the old code is still available at the bottom of the episode (commented out) if you need it. 

One thing I haven't addressed is Sarah's comment about using a dataset not from Tier3. That would mean reimlementing the whole thing from scratch. I would suggest we consider this for a future update.

Peter9192 commented:
Hi @SarahAlidoost and @hb326 I just finished revising the episode. Can you please have another look?

@SarahAlidoost I completely restructured and rephrased everything, so a part of your comments was no longer 1-to-1 applicable. Nevertheless I think I've addressed all of them wherever still possible/relevant. Can you please resolve the conversations as you check my responses? With respect to your main concerns:

- I tried to make everything shorter and changed the structure. I hope it is better now, but I think we cannot avoid that it's still quite a lengthy episode. 
- I removed the custom CMOR table entry (now there's just a comment that this is possible but beyond the scope).
- I agree that it would be nicer to have a variable that's more easy to access/distribute. However,
  - That would mean basically starting again from scratch. Maybe, if we really want it, can we do this in a follow up PR?
  - Somehow it's also helping to make the point that many datasets come with access restrictions.
  - I can imagine we could ask people do download the data in advance whenever we teach this episode

SarahAlidoost commented on _episodes/09-cmorization.md:
````diff
@@ -0,0 +1,772 @@
+```
+
+- **Add a reference**. Make sure that there is a reference file available for
+   the dataset. You can add a [BibTeX info
+   file](https://github.com/ESMValGroup/ESMValTool/tree/master/esmvaltool/references)
+   here. You
````
```suggestion
   the dataset, see the instruction [here](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#adding-references).
```

SarahAlidoost reviewed (approved):
@Peter9192 and @hb326 Thank you for fixing the issues and addressing the comments. It is better-organized and clear now. I approved this PR. Just a minor fix on the text about adding references. Then, let's merge it. :+1:

Peter9192 commented:
Thanks @SarahAlidoost!  @hb326 if you're happy with the lesson as well we can proceed with merging it :-)

hb326 commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,770 @@
+> > filename: 'GPP.ANN.CRUNCEPv6.monthly.*.nc'
+> >
+> > attributes:
+> >   project_id: OBS
```
Should be "OBS6" here, right?

Peter9192 replied:
Well spotted! Changed it

hb326 commented on _episodes/09-cmorization.md:
````diff
@@ -0,0 +1,770 @@
+as it was written with the very basic CMORizer from above.
+
+```bash
+ncdump -h OBS_FLUXCOM_reanaly_ANN-v1_Lmon_gpp_198001-198012.nc
````
And this should be for the year 2000 and not 1980. I had used 1980 as an example before I knew that you, Peter, had instructed people to download the year 2000.

Peter9192 replied:
Check 👍

hb326 commented on _episodes/09-cmorization.md:
```diff
@@ -0,0 +1,770 @@
+
+~~~
+netcdf OBS_FLUXCOM_reanaly_ANN-v1_Lmon_gpp_198001-198012 {
+dimensions:
```
Also here: should be year 2000 and not 1980.

Peter9192 replied:
Done

hb326 commented:
@Peter9192: thanks for your fantastic work on restructuring the episode and going through Sarah's questions and suggestions. I was swamped the last week and could not put any time for the episode development aside. I have just added 3 more comments. Beside these I think the episode is ready to be merged and tested with real-life users!
@SarahAlidoost: thanks so much for your detailed review! The episode is not in much more concise and understandable, I think!
